Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -358,91 +376,120 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
-        if let slice = state.slice {
-            // If we stored a slice before we need to ensure we move the readerIndex so we don't try to parse the data again. We
-            // also need to update the reader index to whatever it is now.
-            state.slice = (buffer.readerIndex, slice.length)
-            buffer.moveReaderIndex(forwardBy: state.readerIndexAdjustment)
+        // Needed to guard again re-entrace calls.
+        self.decoding = true
+        defer {
+            self.decoding = false
         }
 
-        let result = try buffer.withVeryUnsafeBytes { (pointer) -> size_t in
-            state.baseAddress = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+        // Either use the received buffer directly or merge it into the already existing cumulationBuffer.
+        if self.cumulationBuffer == nil {
+            self.cumulationBuffer = buffer
+        } else {
+            self.cumulationBuffer!.write(buffer: &buffer)
+        }
 
-            let result = state.baseAddress!.withMemoryRebound(to: Int8.self, capacity: pointer.count, { p in
-                c_nio_http_parser_execute(&parser, &settings, p.advanced(by: buffer.readerIndex), buffer.readableBytes)
-            })
+        do {
+            repeat {
+                // We need to refetch the cumulationBuffer on each loop as it may has changed due re-entrance calls of channelRead(...)
+                let bufferSlice = self.cumulationBuffer!
 
-            if let error = state.currentError {
-                throw error
-            }
+                let result = try bufferSlice.withVeryUnsafeBytes { (pointer) -> size_t in
+                    state.baseAddress = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+
+                    let result = state.baseAddress!.withMemoryRebound(to: Int8.self, capacity: pointer.count, { p in
+                        c_nio_http_parser_execute(&parser, &settings, p.advanced(by: bufferSlice.readerIndex), bufferSlice.readableBytes)
+                    })
 
-            state.baseAddress = nil
+                    state.baseAddress = nil
 
-            let httpError = parser.http_errno
-            if httpError != 0 {
-                self.state.currentError = HTTPParserError.httpError(fromCHTTPParserErrno: http_errno(rawValue: httpError))!
-                throw self.state.currentError!
+                    if let error = state.currentError {
+                        throw error
+                    }
+
+                    let httpError = parser.http_errno
+                    if httpError != 0 {
+                        self.state.currentError = HTTPParserError.httpError(fromCHTTPParserErrno: http_errno(rawValue: httpError))!
+                        throw self.state.currentError!
+                    }
+                    return result
+                }
+
+                guard self.cumulationBuffer != nil else {
```

Comment:
Would it be better just to not `nil` the buffer out on channel closure? It'll simplify the code here, and the buffer will get dropped anyway, one way or another.

Reply:
I can do this but this would also need changed in `ByteToMessageDecoder` and may increase memory usage if a user still hold a reference to the handler after the channel become inactive / the handler was removed.

WDYT ?